kaltura-common/src/rxjs/operators/cancel-on-destroy.ts

```diff
@@ -0,0 +1,49 @@
+import { Observable, Subject } from 'rxjs';
+import { OnDestroy } from '@angular/core';
+import { merge } from 'rxjs/observable/merge';
+
+interface EnhancedOnDestroy extends OnDestroy {
+    __ngOnDestroySource__: Subject<string>;
+    __ngOnDestroy__: () => void;
+}
+
+export function cancelOnDestroy<T>(instance: OnDestroy, manualDestroy?: Observable<any>): (source: Observable<T>) => Observable<T> {
+    return (source: Observable<T>) => {
+        if (!instance.ngOnDestroy) {
+            return source;
+        }
+
+        return Observable.create(observer => {
+            let subscription = source.subscribe(observer);
+            
+            if (!(<EnhancedOnDestroy>instance).__ngOnDestroySource__) {
+                (<EnhancedOnDestroy>instance).__ngOnDestroySource__ = new Subject();
+                (<EnhancedOnDestroy>instance).__ngOnDestroy__ = instance.ngOnDestroy;
```

Comment:
I think this might be a good place to use Symbol, as I don't think we want this property to be enumerable. plus it will make sure there are no collisions.

Reply:
hmm, never used it before :)
```
const ngOnDestroySource = Symbol('ngOnDestroySource ');
const ngOnDestroy = Symbol('ngOnDestroy ');

interface EnhancedOnDestroy extends OnDestroy {
    [ngOnDestroySource]: Subject<string>;
    [ngOnDestroy]: () => void;
}
//...
if (!instance[ngOnDestroySource]) {
     instance[ngOnDestroySource] = new Subject();
     instance[ngOnDestroy] = instance.ngOnDestroy;
//...
```
something like this, right? (I haven't tried it in the runtime yet, just guessing in the comments)

Reply:
exactly. The symbol data structure is built for use cases like these. I would consult @esakal  first, as I think it requires a polyfill for IE.

Reply:
Hi, It is nice talking about advanced techniques :)

> I think this might be a good place to use Symbol, as I don't think we want this property to be enumerable. plus it will make sure there are no collisions.

It will prevent collisions as you mentioned and will be ignored so it fit the purpose.

> I think it requires a polyfill for IE

We already have this polifill enabled [here](https://github.com/kaltura/kmc-ng/blob/73ebb2176e53900fee0428c959da6b8fc29f3fec/src/polyfills.ts#L22)

Reply:
Okay, apparently it doesn't work yet in typescript v2.5.3 we have in kaltura-ng.
![image](https://user-images.githubusercontent.com/4395889/41521557-4ce92ea4-72db-11e8-92d4-47925a439f63.png)
It relates to this issue https://github.com/Microsoft/TypeScript/issues/5579 and it was closed by this PR https://github.com/Microsoft/TypeScript/pull/15473 which was included into v2.7

@esakal We can leave it as a plain property for now or wait for upgrade of the TS (if it was planned), what do you prefer?